kernel/aster-nix/src/net/socket/ip/datagram/mod.rs

```diff
@@ -296,6 +297,85 @@ impl Socket for DatagramSocket {
         poll_ifaces();
         Ok(sent_bytes)
     }
+
+    fn sendmsg(&self, msg_hdr: MessageHeader, flags: SendRecvFlags) -> Result<usize> {
+        let MessageHeader {
+            addr,
+            io_vecs,
+            control_message,
+        } = msg_hdr;
+
+        if control_message.is_some() {
+            // TODO: support sending control message
+            warn!("sending control message is not supported");
+        }
+
+        let remote_endpoint = match addr {
+            Some(remote_addr) => {
+                let endpoint = remote_addr.try_into()?;
+                self.try_bind_empheral(&endpoint)?;
+                endpoint
+            }
+            None => self.remote_endpoint().ok_or_else(|| {
+                Error::with_message(
+                    Errno::EDESTADDRREQ,
+                    "the destination address is not specified",
+                )
+            })?,
+        };
+
+        let mut total_bytes = 0;
+        for io_vec in IoVecIter::new(&io_vecs) {
+            if io_vec.is_empty() {
+                continue;
+            }
+
+            let buffer = {
+                let mut buffer = vec![0u8; io_vec.len()];
+                io_vec.read_from_user(&mut buffer)?;
+                buffer
+            };
+
+            // TODO: Block if the send buffer is full
+            let sent_bytes = self.try_sendto(&buffer, &remote_endpoint, flags)?;
+            total_bytes += sent_bytes;
+        }
+
+        Ok(total_bytes)
+    }
+
+    fn recvmsg(&self, dst: Box<[IoVec]>, flags: SendRecvFlags) -> Result<(usize, MessageHeader)> {
+        let mut total_bytes = 0;
+        let mut peer_addr = None;
+        for io_vec in IoVecIter::new(&dst) {
+            if io_vec.is_empty() {
+                continue;
+            }
+
+            let (recv_bytes, addr) = {
+                let mut buffer = vec![0u8; io_vec.len()];
+                let (recv_bytes, addr) = self.recvfrom(&mut buffer, flags)?;
```

Comment:
For UDP datagrams, `sendmsg`/`recvmsg` should send/receive only _one_ message. But the current implementation sends/receives multiple messages, which is the behavior of `sendmmsg`/`recvmmsg`, not `sendmsg`/`recvmsg`.